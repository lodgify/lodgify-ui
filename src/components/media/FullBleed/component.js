import React from 'react';
import PropTypes from 'prop-types';
import getClassNames from 'classnames';
import { Segment } from 'semantic-ui-react';
import { size } from 'lodash';

import { ResponsiveImage } from 'media/ResponsiveImage';

import { DEFAULT_BOTTOM_OFFSET } from './constants';

/**
 * A full bleed takes up the width of its containing element
 * and the height of the viewport.
 */
// eslint-disable-next-line jsdoc/require-jsdoc
export const Component = ({
  bottomOffset,
  children,
  hasGradient,
  imageUrl,
  imageHeight,
  imageWidth,
  placeholderImageUrl,
  sizes,
  srcSet,
}) => (
  <Segment
    className={getClassNames('full-bleed', {
      'has-gradient': hasGradient,
      'has-children': size(children) > 0,
    })}
    style={{
      '--full-bleed-bottom-offset': bottomOffset,
    }}
    vertical
  >
    <ResponsiveImage
      imageHeight={imageHeight}
      imageUrl={imageUrl}
      imageWidth={imageWidth}
      isFluid
      placeholderImageUrl={placeholderImageUrl}
      sizes={sizes}
      srcSet={srcSet}
    />
    {children}
  </Segment>
);

Component.displayName = 'FullBleed';

Component.defaultProps = {
  bottomOffset: DEFAULT_BOTTOM_OFFSET,
  children: null,
  hasGradient: false,
  imageHeight: undefined,
  imageUrl: null,
  imageWidth: undefined,
  placeholderImageUrl: null,
  sizes: null,
  srcSet: null,
};

Component.propTypes = {
  /** Reduce the height of the FullBleed with an offset, supports CSS dimensions. */
  bottomOffset: PropTypes.string,
  /** The children to render inside the full bleed. */
  children: PropTypes.node,
  /** Is there a gradient overlaying the full bleed.  */
  hasGradient: PropTypes.bool,
  /** The natural height of the image. */
  imageHeight: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  /** URL pointing to the image to render. */
  imageUrl: PropTypes.string,
  /** The natural width of the image. */
  imageWidth: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  /** URL pointing to the placeholder image to render. */
  placeholderImageUrl: PropTypes.string,
  /** A list of one or more strings separated by commas indicating a set of source sizes for the image. See [the MDN docs for more information](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/img). */
  sizes: PropTypes.string,
  /** A list of one or more strings separated by commas indicating a set of possible image sources for the user agent to use for the image. See [the MDN docs for more information](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/img). */
  srcSet: PropTypes.string,
};
