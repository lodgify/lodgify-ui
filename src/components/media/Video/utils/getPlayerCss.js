/**
 * @param  {boolean}  isResponsive
 * @param  {string}   width
 * @param  {string}   height
 * @return {Object}
 */
export const getPlayerCss = (isResponsive, width, height) =>
  width && height && isResponsive
    ? {
        paddingTop: `${100 / (width / height)}%`,
      }
    : {};
