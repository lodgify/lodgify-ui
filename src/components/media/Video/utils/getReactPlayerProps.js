/**
 * @param  {boolean}  isResponsive
 * @param  {string}   videoSource
 * @return {Object}
 */
export const getReactPlayerProps = (isResponsive, videoSource) => ({
  controls: true,
  className: 'react-player',
  url: videoSource,

  ...(isResponsive
    ? {
        height: '100%',
        width: '100%',
      }
    : {}),
});
