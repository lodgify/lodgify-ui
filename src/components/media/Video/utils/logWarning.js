export const logWarning = videoSource => {
  // eslint-disable-next-line no-console
  console.error(
    `Warning:  \`Video.props.videoSource\` must be a valid URL or valid HTML. Received: ${videoSource}`
  );
};
