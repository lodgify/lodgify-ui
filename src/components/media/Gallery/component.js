import React, { Fragment } from 'react';
import PropTypes from 'prop-types';

import { Modal } from 'elements/Modal';
import { HorizontalGutters } from 'layout/HorizontalGutters';
import { Divider } from 'elements/Divider';
import { Grid } from 'layout/Grid';
import { GridRow } from 'layout/GridRow';
import { GridColumn } from 'layout/GridColumn';
import { buildKeyFromStrings } from 'utils/build-key-from-strings';
import { Heading } from 'typography/Heading';
import { ResponsiveImage } from 'media/ResponsiveImage';

/**
 * A gallery displays a collection of images in a modal.
 */
// eslint-disable-next-line jsdoc/require-jsdoc
export const Component = ({ heading, images, trigger }) => (
  <Modal isFullscreen trigger={trigger}>
    <HorizontalGutters>
      <Divider />
      {heading && (
        <Fragment>
          {heading}
          <Divider hasLine />
        </Fragment>
      )}
      <Grid columns={2} stackable>
        <GridRow>
          {images.map(({ label, url, ...otherProps }, index) => (
            <GridColumn key={buildKeyFromStrings(label, index)}>
              <Heading size="small">{label}</Heading>
              <ResponsiveImage imageUrl={url} {...otherProps} />
              <Divider />
            </GridColumn>
          ))}
        </GridRow>
      </Grid>
      <Divider size="large" />
    </HorizontalGutters>
  </Modal>
);

Component.displayName = 'Gallery';

Component.defaultProps = {
  heading: null,
};

Component.propTypes = {
  /** The content to display as a heading at the top of the gallery. */
  heading: PropTypes.node,
  /** The images to display. See [the `ResponsiveImage` component for valid props](#/Media/ResponsiveImage).  */
  images: PropTypes.arrayOf(PropTypes.object).isRequired,
  /** The element to be clicked to display the gallery. */
  trigger: PropTypes.node.isRequired,
};
