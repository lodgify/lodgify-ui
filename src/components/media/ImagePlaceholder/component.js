import React from 'react';
import { string, oneOfType, number } from 'prop-types';
import classnames from 'classnames';

import { testidFactory } from 'utils/testid';
import { Icon } from 'elements/Icon';

const TEST_PREFIX = 'image-placeholder';
const testid = testidFactory(TEST_PREFIX);

export const Component = ({ className, title, width, height }) => (
  <figure
    className={classnames('image-placeholder', className)}
    data-testid={testid()}
    title={title}
  >
    <Icon name="placeholder" />
  </figure>
);

Component.displayName = 'ImagePlaceholder';
Component.defaultProps = {
  title: null,
  className: '',
};

Component.propTypes = {
  /** Additional ClassName to customize the placeholder. */
  className: string,
  /** The title of the figure that replace the image. */
  title: string
};
