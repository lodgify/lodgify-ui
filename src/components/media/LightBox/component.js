import React from 'react';
import PropTypes from 'prop-types';

import { Modal } from 'elements/Modal';
import { Slideshow } from 'media/Slideshow';
import { testidFactory } from 'utils/testid';

const TEST_ID_PREFIX = 'lightbox';

const testid = testidFactory(TEST_ID_PREFIX);

/**
 * A LightBox which displays an image or a series of images in a modal.
 */
// eslint-disable-next-line jsdoc/require-jsdoc
export const Component = ({ slideShowImages, trigger }) => {
  const areThumbnailsActive = slideShowImages.length > 1;

  return (
    <Modal
      className="is-lightbox"
      data-testid={testid()}
      hasCloseIcon
      isFullscreen
      trigger={trigger}
    >
      <Slideshow
        hasShadow={false}
        images={slideShowImages}
        isRounded={false}
        isShowingBulletNavigation={false}
        isShowingIndex={areThumbnailsActive}
        isShowingThumbnails={areThumbnailsActive}
        thumbnails={slideShowImages}
      />
    </Modal>
  );
};

Component.displayName = 'LightBox';

Component.defaultProps = {
  trigger: null,
};

Component.propTypes = {
  /**  The images to display inside the Slideshow. */
  slideShowImages: PropTypes.arrayOf(
    PropTypes.shape({
      /** A description of the image to show above the slideshow when the image is showing. */
      descriptionText: PropTypes.string,
      /** A list of one or more strings separated by commas indicating a set of source sizes. See [the MDN docs for more information](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/img). */
      sizes: PropTypes.string,
      /** A list of one or more strings separated by commas indicating a set of possible image sources for the user agent to use. See [the MDN docs for more information](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/img). */
      srcSet: PropTypes.string,
      /** Title of the image to show when hovering over it on desktop browsers. */
      title: PropTypes.string,
      /** URL pointing to the image to display. */
      url: PropTypes.string.isRequired,
    })
  ).isRequired,
  /** The element to be clicked to display the modal. */
  trigger: PropTypes.node,
};
