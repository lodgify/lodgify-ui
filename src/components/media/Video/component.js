import ReactPlayer from 'react-player';
import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import DOMPurify from 'dompurify';
import isValidUrl from 'is-url';
import isValidHTML from 'is-html';
import getClassNames from 'classnames';

import { getPlayerCss } from './utils/getPlayerCss';
import { getReactPlayerProps } from './utils/getReactPlayerProps';
import { logWarning } from './utils/logWarning';

/**
 * The Video widget. It allows the consumer to render videos given a URL or
 * an embeddable HTML snippet.
 * @extends {React.PureComponent}
 * @returns {Object}
 */
export class Component extends PureComponent {
  state = {
    cleanHTMLString: null,
  };

  componentDidMount = () => {
    const { videoSource } = this.props;

    isValidHTML(videoSource) &&
      this.setState({
        // `DOMPurify.sanitize` is in `componentDidMount` so that it
        // is not called during server side rendering.
        // Reason: DOMPurify depends on browser features.
        cleanHTMLString: DOMPurify.sanitize(videoSource, {
          SANITIZE_DOM: true,
          ALLOWED_TAGS: ['iframe', 'object', 'video', 'audio'],
        }),
      });
  };

  render() {
    const { videoSource, isResponsive, width, height } = this.props;

    const playerWrapperProps = {
      className: getClassNames('video', 'player-wrapper', {
        'is-url': isValidUrl(videoSource),
        'is-html': isValidHTML(videoSource),
        'is-responsive': isResponsive,
      }),
      style: getPlayerCss(isResponsive, width, height),
    };

    switch (true) {
      case isValidUrl(videoSource):
        return (
          <div {...playerWrapperProps}>
            <ReactPlayer {...getReactPlayerProps(isResponsive, videoSource)} />
          </div>
        );
      case isValidHTML(videoSource):
        return (
          <div
            {...playerWrapperProps}
            dangerouslySetInnerHTML={{
              __html: this.state.cleanHTMLString,
            }}
          />
        );
      default:
        logWarning(videoSource);
        return null;
    }
  }
}

Component.displayName = 'Video';

Component.defaultProps = {
  height: 315,
  isResponsive: false,
  videoSource: null,
  width: 560,
};

Component.propTypes = {
  /** The height of the video to determine the aspect ratio of the video */
  height: PropTypes.number,
  /** Should the video be responsive */
  isResponsive: PropTypes.bool,
  /** The string that will be used to build the video player.
   *  A URL pointing to a Youtube, Vimeo, ... video is valid.
   *  A embeddable HTML video snippet is also valid (iframes, video, audio, objects)
   *  An invalid input throws an error.
   */
  videoSource: PropTypes.string,
  /** The width of the component to determine aspect ratio */
  width: PropTypes.number,
};
