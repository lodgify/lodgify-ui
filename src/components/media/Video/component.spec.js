jest.mock('react-player', () => {
  const Component = props => <div {...props} />;

  Component.displayName = 'ReactPlayer';
  return Component;
});
jest.mock('./utils/logWarning');

import React from 'react';
import { mount } from 'enzyme';
import { expectComponentToHaveDisplayName } from '@lodgify/enzyme-jest-expect-helpers';

import { logWarning } from './utils/logWarning';
import { Component as Video } from './component';

const getVideo = props => mount(<Video {...props} />);

describe('<Video />', () => {
  describe('if `props.videoSource` is a valid URL', () => {
    it('should render the right structure', () => {
      const props = { videoSource: 'https://lodgify.com' };
      const actual = getVideo(props);

      expect(actual).toMatchSnapshot();
    });
  });

  describe('if `props.videoSource` is valid HTML', () => {
    it('should render the right structure', () => {
      const props = { videoSource: '<iframe></iframe>' };
      const actual = getVideo(props);

      expect(actual).toMatchSnapshot();
    });
  });

  describe('if `props.videoSource` is not a valid URL or valid HTML', () => {
    const props = { videoSource: undefined };

    it('should call `logWarning` with the right arguments', () => {
      getVideo(props);

      expect(logWarning).toHaveBeenCalledWith(null);
    });

    it('should render the right structure', () => {
      const actual = getVideo(props);

      expect(actual).toMatchSnapshot();
    });
  });

  describe('if `props.isResponsive` is `true`', () => {
    it('should render the right structure', () => {
      const props = { isResponsive: true };
      const actual = getVideo(props);

      expect(actual).toMatchSnapshot();
    });
  });

  it('should have the right `displayName`', () => {
    expectComponentToHaveDisplayName(Video, 'Video');
  });
});
