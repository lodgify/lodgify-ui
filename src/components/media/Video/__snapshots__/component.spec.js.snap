// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Video /> if \`props.isResponsive\` is \`true\` should render the right structure 1`] = `
<Video
  height={315}
  isResponsive={true}
  videoSource={null}
  width={560}
/>
`;

exports[`<Video /> if \`props.videoSource\` is a valid URL should render the right structure 1`] = `
<Video
  height={315}
  isResponsive={false}
  videoSource="https://lodgify.com"
  width={560}
>
  <div
    className="video player-wrapper is-url"
    style={Object {}}
  >
    <ReactPlayer
      className="react-player"
      controls={true}
      url="https://lodgify.com"
    >
      <div
        className="react-player"
        controls={true}
        url="https://lodgify.com"
      />
    </ReactPlayer>
  </div>
</Video>
`;

exports[`<Video /> if \`props.videoSource\` is not a valid URL or valid HTML should render the right structure 1`] = `
<Video
  height={315}
  isResponsive={false}
  videoSource={null}
  width={560}
/>
`;

exports[`<Video /> if \`props.videoSource\` is valid HTML should render the right structure 1`] = `
<Video
  height={315}
  isResponsive={false}
  videoSource="<iframe></iframe>"
  width={560}
>
  <div
    className="video player-wrapper is-html"
    dangerouslySetInnerHTML={
      Object {
        "__html": "<iframe></iframe>",
      }
    }
    style={Object {}}
  />
</Video>
`;
